src/circuits/simple_gadgets.tcc

```diff
@@ -70,14 +69,22 @@ xor_constant_gadget<FieldT>::xor_constant_gadget(
 template<typename FieldT>
 void xor_constant_gadget<FieldT>::generate_r1cs_constraints()
 {
-    // 32 constraints
+    // Set the constraints (#constraints = length of bit string)
+    //
+    // We know that: res = a XOR b <=> (2.a) * b = a + b - res
+    // we can write: res = a + b - 2ab
+    //
+    // Hence, res2 = a XOR b XOR c = (a XOR b) XOR c <=> res XOR c
+    // Thus: res2 = res XOR c is constrainted as:
+    //   2(res) * c = res + c - res2
+    // which leads to:
+    //   2(a + b - 2ab) * c = a + b - 2ab + c - res2
     for (size_t i = 0; i < a.size(); i++) {
         this->pb.add_r1cs_constraint(
             libsnark::r1cs_constraint<FieldT>(
-                -FieldT("2") * (FieldT("1") - FieldT("2") * c[i]) * a[i],
-                b[i],
-                res[i] - c[i] - a[i] * (FieldT("1") - FieldT("2") * c[i]) -
-                    b[i] * (FieldT("1") - FieldT("2") * c[i])),
+                2 * (a[i] + b[i] - 2 * a[i] * b[i]),
+                c[i],
+                a[i] + b[i] - 2 * a[i] * b[i] + c[i] - res[i]),
```

Comment:
The variable c was supposed to be the constants which is the reason why I wrote the formula like this. It seems here that you consider b to be the constant.
I supposed we should write what variable is the constant somewhere in the code not to confuse the reader.